MekHQ/src/mekhq/campaign/autoresolve/acar/handler/StandardUnitAttackHandler.java

```diff
@@ -0,0 +1,231 @@
+        if (criticalRollResult <= 4) {
+            // No crit
+            reporter.reportNoCrit();
+        } else if (criticalRollResult <= 7) {
+            // targeting critical
+            targetUnit.addTargetingCrit();
+            reporter.reportTargetingCrit(targetUnit);
+        } else if (criticalRollResult <= 9) {
+            // Damage crit
+            targetUnit.addDamageCrit();
+            reporter.reportDamageCrit(targetUnit);
+        } else if (criticalRollResult <= 11) {
+            // Both targeting and damage crit
+            targetUnit.addTargetingCrit();
+            targetUnit.addDamageCrit();
+            reporter.reportTargetingCrit(targetUnit);
+            reporter.reportDamageCrit(targetUnit);
+        } else {
+            // Destroyed
+            countKill(attackingUnit, targetUnit);
+            targetUnit.setCurrentArmor(0);
+            target.setHighStressEpisode();
+            reporter.reportUnitDestroyed();
+        }
+    }
+
+
+    private void countKill(SBFUnit attackingUnit, SBFUnit targetUnit) {
+
+        var killers = attackingUnit.getElements().stream().map(AlphaStrikeElement::getId)
+            .map(e -> simulationManager().getGame().getEntity(e)).filter(Optional::isPresent).map(Optional::get).toList();
+        var targets = targetUnit.getElements().stream().map(AlphaStrikeElement::getId)
+            .map(e -> simulationManager().getGame().getEntity(e)).filter(Optional::isPresent).map(Optional::get).toList();
+        for (var target : targets) {
+            RandomUtils.fastSample(killers).ifPresent(e -> e.addKill(target));
+        }
+    }
+
+    private int processDamageByEngagementControl(Formation attacker, Formation target, int damage) {
+        var engagementControlMemories = attacker.getMemory().getMemories("engagementControl");
+        var engagement = engagementControlMemories
+            .stream()
+            .filter(f -> f.getOrDefault("targetFormationId", Entity.NONE).equals(target.getId()))
+            .findFirst(); // there should be only one engagement control memory for this target
+
+        if (engagement.isEmpty()) {
+            return damage;
+        }
+
+        var isAttacker = (boolean) engagement.get().getOrDefault("attacker", false);
+
+        if (!isAttacker) {
+            return damage;
+        }
+
+        var wonEngagement = (boolean) engagement.get().getOrDefault("wonEngagementControl", false);
+        var engagementControl = (EngagementControl) engagement.get().getOrDefault("engagementControl", EngagementControl.NONE);
+        if (wonEngagement) {
+            return processDamageEngagementControlVictory(engagementControl, damage);
+        }
+        return processDamageEngagementControlDefeat(engagementControl, damage);
+    }
+
+    private int processDamageEngagementControlDefeat(EngagementControl engagementControl, double damage) {
+        switch(engagementControl) {
```

Comment:
We probably want to be using enhanced switches here, unless you want specific fall through behavior, which it doesn't look like you do.

Enhanced switches only became available to us when we updated to j17, so you'll only see it in newer code.

Reply:
fixing